kv/storage/standalone_storage/standalone_storage.go

```diff
@@ -1,38 +1,107 @@
 package standalone_storage
 
 import (
+	"log"
+	"os"
+	"sync"
+
+	"github.com/Connor1996/badger"
 	"github.com/pingcap-incubator/tinykv/kv/config"
 	"github.com/pingcap-incubator/tinykv/kv/storage"
+	"github.com/pingcap-incubator/tinykv/kv/util/engine_util"
 	"github.com/pingcap-incubator/tinykv/proto/pkg/kvrpcpb"
 )
 
 // StandAloneStorage is an implementation of `Storage` for a single-node TinyKV instance. It does not
 // communicate with other nodes and all data is stored locally.
 type StandAloneStorage struct {
 	// Your Data Here (1).
+	//storage engine
+	db     *badger.DB
+	config *config.Config
+	once   *sync.Once
 }
 
 func NewStandAloneStorage(conf *config.Config) *StandAloneStorage {
 	// Your Code Here (1).
-	return nil
+	//new a StandAloneStorage DB
+	return &StandAloneStorage{
+		config: conf,
+		once:   &sync.Once{},
+	}
 }
 
+//Start
 func (s *StandAloneStorage) Start() error {
-	// Your Code Here (1).
-	return nil
+	DBPath := s.config.DBPath
+	_, err := os.Stat(DBPath)
+	if err != nil {
+		if os.IsExist(err) {
+			log.Fatalf("standard alone sotrage start failed :%s", err)
+		}
+		s.db = engine_util.CreateDB(DBPath, false)
+		return nil
+	}
+	s.db = engine_util.CreateDB(DBPath, false)
+	return err
 }
 
+//Stop
 func (s *StandAloneStorage) Stop() error {
-	// Your Code Here (1).
+	s.once.Do(func() {
+		s.db.Close()
+	})
 	return nil
 }
 
 func (s *StandAloneStorage) Reader(ctx *kvrpcpb.Context) (storage.StorageReader, error) {
 	// Your Code Here (1).
-	return nil, nil
+	return &StandAloneStorageReader{
+		db: s.db,
+	}, nil
 }
 
 func (s *StandAloneStorage) Write(ctx *kvrpcpb.Context, batch []storage.Modify) error {
 	// Your Code Here (1).
-	return nil
+	wb := &engine_util.WriteBatch{}
+	for _, v := range batch {
+		switch v.Value() {
+		case nil:
+			wb.DeleteCF(v.Cf(), v.Key())
+		default:
+			wb.SetCF(v.Cf(), v.Key(), v.Value())
+		}
+	}
+	err := wb.WriteToDB(s.db)
+	return err
+}
+
+var _ storage.StorageReader = new(StandAloneStorageReader)
+
+type StandAloneStorageReader struct {
+	db *badger.DB
+}
+
+//GetCF
+func (s *StandAloneStorageReader) GetCF(cf string, key []byte) ([]byte, error) {
+	txn := s.db.NewTransaction(false)
+	defer txn.Discard()
+	val, err := engine_util.GetCFFromTxn(txn, cf, key)
+	if err != nil {
+		if err.Error() == badger.ErrKeyNotFound.Error() {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return val, nil
+}
+
+//IterCF
+func (s *StandAloneStorageReader) IterCF(cf string) engine_util.DBIterator {
+	txn := s.db.NewTransaction(false)
+	return engine_util.NewCFIterator(cf, txn)
```

Comment:
Do we need to consider when txn is discarded?

Reply:
Maybe reader txn do not need？ i m not sure